doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+
+    /**
+     * @brief Set number of frame retransmissions
+     *
+     * @pre the device is not sleeping
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       frame retransmissions
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] retries the number of retransmissions
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_frame_retries)(ieee802154_dev_t *dev, int retries);
+
+    /**
+     * @brief Set IEEE802.15.4 CSMA configuration parameters
+     *
+     * @pre the device is not sleeping
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       frame retransmissions
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] bd parameters of the exponential backoff
+     * @param[in] retries number of CSMA-CA retries. If @p retries < 0, 
+     *                    retransmissions with CSMA-CA MUST be disabled.
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_csma_params)(ieee802154_dev_t *dev, ieee802154_csma_be_t *bd,
+                           int8_t retries);
+```
+
+## 6.2 Event Notification
+
+The Event Notification mechanism is implemented with a function callback.
+The callback function is supposed to be implemented by the upper layer.
+
+The callback signature, the events and their expected behavior are defined
+in the following block:
+
+```c
+/**
+ * @brief   Event notification status
+ */
+typedef enum {
+    /**
+     * @brief the transceiver detected a valid Start Frame Delimiter, which
+              indicates the radio started to receive a packet.
+     */
+    IEEE802154_RADIO_RX_START,
+
+    /**
+     * @brief the transceiver received a packet and there is a packet in the
+     *        internal framebuffer.
+     *
+     * The transceiver is in @ref IEEE802154_TRX_STATE_RX_ON state when
+     * this function is called, but with framebuffer protection (either
+     * dynamic framebuffer protection or disabled RX). Thus, the packet
+     * will not be overwritten before calling the @ref ieee802154_radio_read
+     * function. However, @ref ieee802154_radio_read MUST be called in
+     * order to receive new packets. If there is no interest in the
+     * packet, the function can be called with a NULL buffer to drop
+     * the packet.
+     */
+    IEEE802154_RADIO_RX_DONE,
+
+    /**
+     * @brief the transceiver finished sending a packet or the
+     *        retransmission procedure.
+     *
+     * If the radio supports frame retransmissions the
+     * @ref ieee802154_radio_get_tx_status MAY be called to retrieve useful
+     * information (number of retries, frame pending bit, etc). The
+     * transceiver is in @ref IEEE802154_TRX_STATE_TX_ON state when this function
+     * is called.
+     */
+    IEEE802154_RADIO_TX_DONE,
+} ieee802154_trx_status_t;
```

Comment:
I think the events can be generic to all wireless network devices.

Additionally, as already said above: I would like this API to allow non-blocking use. As a result, more events are needed. E.g. in a blocking mode you can simply do:

```C
prepare();
transmit();
```

But if the `prepare()` does not block, the `transmit()` might return `-EAGAIN` (or `-EBUSY`?) when the `prepare()` has not yet completed. (`prepare()` could take some significant time if the device has the frequency generator offline.)

Instead, an event would need to be triggered when the device is ready to transmit. One could add some convenience code to block until an event is triggered, e.g. it could be used like this:

```
prepare();
await_event(READY_FOR_TX);
transmit();
```

Reply:
The only thing that I'm concerned here: if we need to care about "blocking" stuff here at radio level, IMO a higher layer is not doing it's job well.


![file](https://user-images.githubusercontent.com/1260616/80313597-1c985b80-87ec-11ea-90c8-7ff2b2ea0606.png)

By design, these Wireless MAC layers work under a request/confirm architecture (MLME-SAP, MCPS-SAP) which are by nature non blocking. A user "request" the MAC layer to send data, and the MAC layer might respond with a "queue is full" or "queued". 

So, the MAC is the non-blocking layer in the IEEE802.15.4 architecture.

Also, consider the IEEE802.15.4 PHY definition:
![file](https://user-images.githubusercontent.com/1260616/80313696-ead3c480-87ec-11ea-9a51-82a8d0769100.png)


Basically the Radio HAL is the RF-SAP in the IEEE802.15.4 PHY model. The specification doesn't say anything about the blocking/non-blocking nature of the PD-SAP interface, but IMO the interface definition suggest that this is not the layer that should care about this.
As an example, this is the "confirm" primitive of the PD-DATA.request command (used to send data at PHY level):

![file](https://user-images.githubusercontent.com/1260616/80313912-5ec29c80-87ee-11ea-9e2a-50da098f5e94.png)

Another example, some APIs (OpenWSN) expect the radio to simply transmit data (because the media access is already controlled by the MAC). In fact, the "transmit" function is a `void` function and the stack has no way to deal with a -EBUSY (if it happens for whatever reason)

Reply:
> The only thing that I'm concerned here: if we need to care about "blocking" stuff here at radio level, IMO a higher layer is not doing it's job well.

Well, the current situation is that e.g. GNRC happily calls `send()` at every given point in time no matter if the device is currently already sending a frame or not. If the driver does not block during `send()` and returns an `-EBUSY`, the frame is just lost. As a result when e.g. using 6LoWPAN and sending any packet needing layer 2 fragmentation, all but the first fragment are lost.

> In fact, the "transmit" function is a void function and the stack has no way to deal with a `-EBUSY` (if it happens for whatever reason)

This would not be an issue, if the upper layer waits for the driver to be ready for transmission first. E.g. in the state machine above, a transition to the state "Ready for TX" would trigger an event the MAC could wait for / react to. And in that state, a `transmit()` never fails.

So, if the MAC layer synchronizes with the driver, an `-EBUSY` should never ever happen.

Reply:
> Well, the current situation is that e.g. GNRC happily calls send() at every given point in time no matter if the device is currently already sending a frame or not. If the driver does not block during send() and returns an -EBUSY, the frame is just lost. As a result when e.g. using 6LoWPAN and sending any packet needing layer 2 fragmentation, all but the first fragment are lost.

I strongly agree with this. But the actual problem here is the lack of MAC layers, not the implementation of the send function. In practice, GNRC would interact with the MAC layer via non blocking requests (it becomes even more important when using slotted MACs, unless we want to block until the slot is available)

> So, if the MAC layer synchronizes with the driver, an -EBUSY should never ever happen.

That's my point. If the radio HAL provides resources to know when it's safe to send again (event callbacks, etc), why do we need to add an -EBUSY here?

Reply:
> That's my point. If the radio HAL provides resources to know when it's safe to send again (event callbacks, etc), why do we need to add an -EBUSY here?

Sorry for being unclear, the `-EBUSY` was not my point. I would only like the Radio HAL to define that a call to `transmit()` fails, unless `prepare()` has been called previously and the device signaled completion of the `prepare()`. I don't mind if the error is then reported using a return value, or an `assert()` blows, or a log message is generation, or the frame is silently dropped. I really just want the `transmit()` to fail when called in an inappropriate state.

Reply:
I understand the intention here, but under what circumstance you would get to a state where you are not sure if it's safe to call "transmit"? With this API, when the "prepare" function finishes the radio is ready to transmit (FB is already loaded). The only case I could imagine is the case where we use some kind of DMA to prepare a frame. But with the current "iolist" this would be hard anyway (I'm not aware of scattering/gathering DMA APIs).

If a MAC layer decides to prepare in a previous cycle and transmit later, the API can in practice rely on the TX_DONE event.

I just did some measurements for the samr21-xpro (basic mode, direct send). For sending 127 bytes packet, it takes around 400 us to "prepare" the frame and ~4300 us from TX_START to TX_END (TOA). Considering the LIFS is 640us, calling "prepare" right after the TX_DONE event is still compliant with sending packets with IFS >= 640us.

Maybe we could add a precondition that the "prepare" function should be called while there's no ongoing transmission (e.g after TX_DONE)?

Reply:
Is it allowed to call `prepare()` when the transceiver state is not TX? When `prepare()` really just copies over the buffer and does not need to wait for the PLL to settle, `prepare()` could block. (As said, to me blocking for an SPI transfer does count as "non-blocking" :wink:) In that case, there is no need for an event to notify the HAL user that `prepare()` finished.

However, a simplistic user of the HAL (e.g. an ALOHA MAC) would still have to await an event for calling `prepare()` again after having issued `transmit()`. For simplicity, this could just be the TX_COMPLETED event.

Reply:
> Is it allowed to call prepare() when the transceiver state is not TX?

Yes! In fact, it is preferred to call `prepare` during TRX_OFF (if there are no timing constraints, considering switching from RX_ON to TX_ON takes 1 us but from TRX_OFF ~200 us). TRX_OFF consumes ~0.5 mA in while TX_ON consume ~6 mA (`at86rf2xx`).

> However, a simplistic user of the HAL (e.g. an ALOHA MAC) would still have to await an event for calling prepare() again after having issued transmit(). For simplicity, this could just be the TX_COMPLETED event.

Exactly, that's the idea